rfcs/0010-field-arguments.md

```diff
@@ -0,0 +1,57 @@
+# Field Arguments
+
+The ability to provide arguments to fields broadens the use-cases that NDC can support.
+
+This can be seen as generalising `fields` to `functions` or, simply as adding context to fields.
+
+## Proposal
+
+This proposes an update to the NDC-Spec to allow arguments to fields.
+
+The motivation for this change is to generalise the API surface to allow more expressive queries - This should be very similar to collection arguments in practice but can apply to any field (especially nested), not just top-level collections.
+
+Some examples of why this might be useful:
+
+* JSON Operations: For a JSON field in a PG table, several JSON functions could be exposed as fields but they require arguments to be useful. Such as [#>](https://www.postgresql.org/docs/9.3/functions-json.html)
+* Vector Operations for LLMs etc.
+* Advanced Geo Operations
+* GraphQL federation: Forwarding Hasura GQL schemas over NDC will require this change if we want to expose root fields as commands instead of collections.
+
+The schema is extended with:
+
+```rust
+pub struct ObjectField {
+  ...
+  pub arguments: Option<BTreeMap<String, ArgumentInfo>>,
+}
+```
+
+and queries are extended with:
+
+```rust
+pub enum Field {
+  Column {
+    ...
+    arguments: Option<BTreeMap<String, Argument>>,
+  }
+  ...
+}
+```
+
+This mirrors the existing implementation for collection arguments with the
+addition of an `Option` wrapper in order to make the implementation backwards
+compatible. RFC for this aspect - Should these be non-optional?
+
+## Implications
+
+NDC schema and query invocation:
+
+* When the schema indicates that a field has arguments then they may be provided in a query.
```

Comment:
Let's please take the same approach as other types of arguments and say that all arguments are required whether nullable or not.

Reply:
Couldn't this bloat the payload significantly?

Reply:
Especially as we start to leverage common field arguments such as limit, etc. A query which has no field constraints could then require every argument to be specified as null explicitly.

Reply:
I guess we could consider this an optimisation concern and address later along with other similar issues.

Reply:
Yes it's probably not necessary, but let's be consistent for now, and we can relax this later for all types of nullable input arguments at once.

Reply:
👍